### ISSUE TITLE
Semantic database

### ISSUE THREAD

**Issue by xeno-by** [scalameta/scalameta#605]
A **semantic database** is a storage for platform-independent semantic information about Scala code. It is a critical component of the current approach to the scala.meta semantic API (#604).

Previously, we planned to datamine that information from scalac internal trees and store it in platform-independent ASTs. However, as practice has shown (#148), the associated engineering complexity of such an approach is prohibitively high.

At the moment, we're exploring a simpler, more low-level approach at obtaining and persisting semantic information. This approach works similarly to what IDEs do - for every relevant position in Scala sources, we provide a way to query relevant semantic information pertaining to it.

This ticket is dedicated to discussing and specifying the associated data structures, their persistence format, as well as possible use cases. What you're seeing here is a work in progress that may change as we invest more effort into the implementation.

Data structures
===========

```scala
case class Location(uri: String, start: Int, end: Int)
```

**Location** is a range position in a given snippet of Scala code. It is reminiscent of the `Location`/`Range`/`Position` trio from the [Language Server Protocol](https://github.com/Microsoft/language-server-protocol) combined into a single data structure. 

At the moment, URI format is implementation-dependent, but we'll need to become more formal as the implementation evolves (also consider #442 here).

```scala
case class Symbol(id: String)
```

**Symbol** is an extremely lightweight wrapper over a UID of a definition. The UID is currently modelled as a string to simplify prototyping and experimenting, and is inspired by [Symbol](https://github.com/scalameta/scalameta/blob/v1.5.0/scalameta/trees/src/main/scala/scala/meta/internal/semantic/Denotation.scala#L41) from Scala.meta 1.x. Concretely,
  * Local symbols are represented as `<uri>@<start>..<end>` (where `start` is inclusive and `end` is non-inclusive) , e.g. `file:/Users/xeno_by/Projects/nirvana/sandbox/Test.scala@2192..2195`.
  * Global symbols are represented as strings that are similar to fully-qualified names, with every symbol in an owner chain encoded using its name and designator, e.g. `_root_.scala.Int#` or `_root_.geny.Generator#generate(Lscala/Function1;)Lgeny/Generator/Action;.(handleItem)`. More concretely:
    * Types are encoded as `<name>#`
    * Terms are encoded as `<name>.`
    * Methods are encoded as `<name><jvmsig>.`
    * Type parameters are encoded as `[<name>]`
    * Term parameters are encoded as `(<name>)`
    * Self type annotations are encoded as `<name>=>`
    * Non-alphanumeric names are wrapped in backticks

```scala
case class Database(
  symbols: Map[Location, Symbol],
  // To be discussed
)
```

**Semantic database** consists in:
  * A map that, for every `scala.meta.Name` in a corresponding Scala source or sources, contains a `Symbol` corresponding to the name's `Location`.
  * To be discussed 

Persistence format
==============

Currently, semantic databases are created on per-project basis and persisted at the root of the output directory in a text file called `semanticdb`. The text format is extremely naive, but pleasant for eyeballing and amenable to shell scripting. Here's how a semantic database looks like for a simple Scala program:

```scala
object Test {
  def printList(): Unit = {
    println(List(1, 2, 3))
  }
}
```

```
file:/Users/xeno_by/Projects/sbt-semantic-example/library/src/main/scala/Test.scala
[7..11): Test => _empty_.Test.
[20..29): printList => _empty_.Test.printList()V.
[33..37): Unit => _root_.scala.Unit#
[46..53): println => _root_.scala.Predef.println(Ljava/lang/Object;)V.
[54..58): List => _root_.scala.collection.immutable.List.apply(Lscala/collection/Seq;)Lscala/collection/immutable/List;.
```

We're planning to make several improvements to semantic db persistence. First of all, we want to use a robust binary format to serialize semantic databases (#695). Moreover, we'd like to split a monolithic `semanticdb` into a bunch of fine-grained databases (#666).

**Comment by xeno-by** [comment]
@fommil What do you think about this? I heard that you have something similar working in Ensime. /cc @smarter

**Comment by fommil** [comment]
We don't really have a semantic database, but we have a database of symbols and types (which will be in a graph database shortly) and we attach information there such as source code location and (soon) bytecode <=> java <=> scala name conversion. We also index it with Lucene to allow for camel case and fuzzy search.

This is built up from **binary** data, not source code. We let the PC deal with providing this information if it is available in the currently open files. We are looking in the raw bytecode and also augmenting it with some scalap information (I'd love it if future versions of scala had an alternative to the limited, and difficult to use, scalap output)

What you're suggesting would be excellent meta data to store in the classfiles.

For large projects, it can take a long time to build up the index and we've optimised the hell out of it.

Note that including full URIs is not transferable between machines, and will break reproducible builds (scalactic makes the full path an option to facilitate reproducible builds from different dirs).

**Comment by ShaneDelmore** [comment]
@fommil Storing that metadata is a wonderful idea.  Speaking of graph queries, longer term are there plans to be able to run graph type queries against the semantic database?

It would be useful to be able to ask questions such 

- How many callers are there of this method?  Is it dead code?
- Do external callers exist, or could this method be made private?
- How many calls are made into a.b.c namespace?  Am I pulling in a whole library for one function?
- Do I have cyclic dependencies?
- What calls would I need to change to be able to extract this package into a standalone module.

**Comment by ShaneDelmore** [comment]
Will the semantic database be able to be queried bi-directionally?  
For example, instead of having to traverse all of my code asking the db for types to identify all usages of scala.concurrent.future, it would be useful to be able to query the db directly for all code locations or trees of type scala.concurrent.future and apply a transform to them.

**Comment by fommil** [comment]
We have a few tickets open to do these sorts of things, but I don't think anybody is actively looking at implementing them yet. Once the graph PR is merged, we can do a call to action.

**Comment by fommil** [comment]
Have a look at indexer tagged tickets and create a new one for each of these ideas, then whoever is interested can discuss it more there. I think rory was keen on looking at dead code discovery.

**Comment by ShaneDelmore** [comment]
@fommil I am not clear on the relationship between ensime and scalameta. Are you advocating implementing some features in ensime instead of scalameta?  Or having ensime depend on scalameta?

**Comment by fommil** [comment]
I'm talking just ensime here. Scalameta won't let you do everything you're thinking of here.

**Comment by xeno-by** [comment]
@fommil Do you see a way for us to collaborate in order to avoid reimplementing each other's work? The data structures and data formats are different, for sure, but maybe there could still be possibility for sharing?

@ShaneDelmore Regarding your question on whether the semantic database will be able to be queried bi-directionally. We can arrange that, I think. Now that we have an implementation (https://github.com/scalameta/scalameta/pull/636#issuecomment-276843014), we can start playing with possibilities. Wdyt?

**Comment by fommil** [comment]
@xeno-by it would bring in a dependency on shapeless and orientdb, I'm not sure you really want to do that. There is a lot more value in scala.meta providing a rich metadata structure in `.class` files and then allowing tooling (such as ensime) to built on top of that... so an opt-in replacement / complement to scalap.

I think I'd need to understand more about what you want to provide in the API, and some use cases.

**Comment by fommil** [comment]
@ShaneDelmore @xeno-by I thought the question was at ensime... if its db could be queried

**Comment by xeno-by** [comment]
@fommil So far, the goal is to store symbols for all idents. Further down the road, I'd also like to store types for all terms. Storing inferred type and term arguments also seems desirable, but I'm not sure how exactly we should approach that.

The use cases are dictated by the needs of the semantic API. We'd like to provide our users with ways to resolve names, calculate types, list members, etc.

Would this kind of information be useful to you? Anything you'd like to see added?

**Comment by fommil** [comment]
It definitely is useful. But i don't think you need to provide the global lookup mechanism. Per class storage is fine.

This is limited to files that have been parsed by scala.meta, so won't work for deps and java. Hence I think it makes most sense for compile time.

**Comment by fommil** [comment]
btw, my scalasphere talk is going to be on how ensime's indexer works, so hopefully you'll have a better idea of what it will do then. My preference for this semantic database would be if scala.meta creates the data (and provides a mechanism to deserialise it), and then an abstract API that is to be implemented downstream. Then scala-ide and ensime can provide their own implementations based on their indexers, and you could maybe release a basic reference impl if you like, e.g. for integration testing on small projects.

**Comment by fommil** [comment]
(incidentally, scala-ide and ensime devs have spoken about this kind of architecture before for the presentation compiler... but neither of us has had the time)

**Comment by xeno-by** [comment]
@fommil This sounds very interesting. Talk to you in Krakow!

**Comment by xeno-by** [comment]
@densh suggested that we may be better off with less concise, but also less cryptic symbol ids. For example, instead of `scala.Int#`, we could say `scala : type Int` or something similar.

**Comment by xeno-by** [comment]
This ticket was created to provide a specification describing the format of semantic databases + a forum to discuss features that should be supported. I planned to update the ticket as our implementation evolved and engage in discussions from the community.

I think that by now both of this goals have already been achieved.

First, for quite a while now we have a protobuf schema that describes the format of semantic databases. Here's a link to the schema for the latest stable release at the time of writing: https://github.com/scalameta/scalameta/blob/v1.7.0/scalameta/semantic/shared/src/main/protobuf/semanticdb.proto. I propose we treat this schema as the single source of truth regarding the format.

Secondly, the discussion about desired features of semantic databases has concluded. It is clear that we want to have information about big things like resolved names, inferred types and implicit arguments. More ideas may come up in the future, but we can discuss them in dedicated issues.

Therefore, I'll be closing this ticket now. Thank you, everyone, for your input!

**Comment by fommil** [comment]
Cool!

**Comment by Atry** [comment]
> This is built up from binary data, not source code. 
Does it mean all type parameters and dependent types are erased in `Mirror`?

**Comment by xeno-by** [comment]
`Mirror` is designed to reflect code exactly as it's been written - with all the wealth of the Scala type system available for inspection.

We don't currently store all types in the semantic db (still need to implement that), but what we store and what we will be storing is going to be unerased.

**Comment by Atry** [comment]
This looks like erased to me:
`[54..58): List => _root_.scala.collection.immutable.List.apply(Lscala/collection/Seq;)Lscala/collection/immutable/List;.`

**Comment by xeno-by** [comment]
What you see to the right of `=>` is a symbol id, which for methods is built from a JVM signature of the method, as per the Scala Language Specification. We intend for symbols ids to be opaque tokens, so their exact shape shouldn't be representative of the capabilities of our semantic API.

On the other hand, if you try to build a semantic db with the recently released 1.8.0, you'll see that denotations contain unerased type signatures. That's the level of detail that we're aiming to provide to our users.
